cni/pkg/taint/taintcontroller.go

```diff
@@ -0,0 +1,414 @@
+	}
+	return nil
+}
+
+//controller will run all of the critical pod controllers and node controllers, process node and pod in every second
+func (tc *Controller) Run(stopCh <-chan struct{}) {
+	for _, podcontroller := range tc.podController {
+		go podcontroller.Run(stopCh)
+		if !cache.WaitForCacheSync(stopCh, podcontroller.HasSynced) {
+			runtime.HandleError(fmt.Errorf("timed out waiting for caches to sync pod controller"))
+			return
+		}
+	}
+	go tc.nodeController.Run(stopCh)
+	if !cache.WaitForCacheSync(stopCh, tc.nodeController.HasSynced) {
+		runtime.HandleError(fmt.Errorf("timed out waiting for caches to sync node controller"))
+		return
+	}
+	// This will run the func every 1 second until stopCh is sent
+	go wait.Until(
+		func() {
+			// Runs processNextItem in a loop, if it returns false it will
+			// be restarted by wait.Until unless stopCh is sent.
+			for tc.processNextPod() {
+			}
+		},
+		time.Second,
+		stopCh)
+	// This will run the func every 1 second until stopCh is sent
+	go wait.Until(
+		func() {
+			// Runs processNextItem in a loop, if it returns false it will
+			// be restarted by wait.Until unless stopCh is sent.
+			for tc.processNextNode() {
+			}
+		},
+		time.Second,
+		stopCh)
+	<-stopCh
+	log.Infof("stop taint readiness controller")
+}
+func (tc *Controller) processNextPod() bool {
+	errorHandler := func(obj interface{}, pod *v1.Pod, err error) {
+		if err == nil {
+			log.Debugf("Removing %s/%s from work queue", pod.Namespace, pod.Name)
+			tc.podWorkQueue.Forget(obj)
+		} else if tc.podWorkQueue.NumRequeues(obj) < 50 {
+			log.Errorf("Error: %s", err.Error())
+			log.Infof("Re-adding %s/%s to work queue", pod.Namespace, pod.Name)
+			tc.podWorkQueue.AddRateLimited(obj)
+		} else {
+			log.Infof("Requeue limit reached, removing %s/%s", pod.Namespace, pod.Name)
+			tc.podWorkQueue.Forget(obj)
+			runtime.HandleError(err)
+		}
+	}
+	obj, quit := tc.podWorkQueue.Get()
+	if quit {
+		// Exit permanently
+		return false
+	}
+	defer tc.podWorkQueue.Done(obj)
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		log.Errorf("Error decoding object, invalid type. Dropping.")
+		tc.podWorkQueue.Forget(obj)
+		// Short-circuit on this item, but return true to keep
+		// processing.
+		return true
+	}
+	//check for pod readiness
+	//if pod is ready, check all of critical labels in the corresponding node, if all of them are ready, untaint the node
+	//if pod is not ready or some critical pods are not ready, taint the node
+	if podutils.IsPodReady(pod) {
+		err := tc.processReadyPod(pod)
+		errorHandler(obj, pod, err)
+	} else {
+		err := tc.processUnReadyPod(pod)
+		errorHandler(obj, pod, err)
+	}
+	// Return true to let the loop process the next item
+	return true
+}
+
+//list all candidate pods given node name, namespace and selector using cached storage
+func (tc Controller) listCandidatePods(nodeName string, namespace string, selector string) []*v1.Pod {
+	if _, ok := tc.cachedPodsStore[namespace]; !ok {
+		return []*v1.Pod{}
```

Comment:
nit: can just return nil

Reply:
also you can merge this with L234, but doesn't really matter. lookups into nil maps are ok in golang

Reply:
return nil for both of them